builtin/providers/aws/resource_aws_security_group.go

```diff
@@ -313,34 +316,49 @@ func resourceAwsSecurityGroupRuleHash(v interface{}) int {
 	return hashcode.String(buf.String())
 }
 
-func resourceAwsSecurityGroupIPPermGather(d *schema.ResourceData, permissions []ec2.IPPerm) []map[string]interface{} {
+func resourceAwsSecurityGroupIPPermGather(d *schema.ResourceData, permissions []ec2.IPPermission) []map[string]interface{} {
 	ruleMap := make(map[string]map[string]interface{})
 	for _, perm := range permissions {
-		k := fmt.Sprintf("%s-%d-%d", perm.Protocol, perm.FromPort, perm.ToPort)
+		var fromPort, toPort int
+		if perm.FromPort == nil {
+			fromPort = 0
+		} else {
+			fromPort = *perm.FromPort
+		}
+		if perm.ToPort == nil {
+			toPort = 0
+		} else {
+			toPort = *perm.ToPort
+		}
```

Comment:
Lifechanging edit:

```go
var fromPort, toPort int
if v := perm.FromPort; v != nil {
    fromPort = *v
}
```

(same for `toPort`).

The zero value for ints is zero, so you don't need to set it to zero. This is specified in the Go language specification so it is safe to code this way.